design/deploy-steps.md

```diff
@@ -0,0 +1,137 @@
+<!--
+ This work is licensed under a Creative Commons Attribution 3.0
+ Unported License.
+
+ http://creativecommons.org/licenses/by/3.0/legalcode
+-->
+
+# Customizable deployment procedure
+
+## Status
+
+implementable
+
+## Summary
+
+This design proposes customizing the deployment procedure by providing a high
+level abstraction around Ironic [deploy steps][deploy-steps], as well as
+allowing per-host deployment (IPA) images.
+
+[deploy-steps]: https://docs.openstack.org/ironic/latest/admin/node-deployment.html#deploy-steps
+
+## Motivation
+
+By default the Ironic agent downloads a provided image via HTTP(s), converts it
+(if needed) and writes directly to the disk. This procedure works for many
+distributions, but may not suit some cases well. A prominent example is CoreOS,
+which has its own installer (`coreos-installer`) that allows more
+customization.
+
+### Goals
+
+- A way to execute custom actions during deployment.
+- A way to replace the default deploy procedure with the custom actions.
+- A way to provide a link to the Ironic agent image per host.
+
+### Non-Goals
+
+- Expose the full functionality of Ironic deploy steps.
+- Building custom agent images.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a CoreOS user I want to use `coreos-installer` to write the CoreOS image
+without losing any functionality that Ironic offers (unlike the live ISO
+workflow).
+
+#### Story 2
+
+As an operator I want to execute site-specific actions during deployment. I'm
+ready to provide them as plugins to the Ironic agent inside my custom agent
+image.
+
+## Design Details
+
+Additions to `BareMetalHost`:
+
+- New structure field `AgentImage` with the only field `URL`. It will accept
+  an optional link to a custom agent ISO image.
+
+NOTE: custom kernel/initramfs are out of scope for this proposal.
+
+- New structure field `CustomDeploy` with two field `DefaultSteps` and `Steps`.
+- `Steps` is a list of deploy steps to apply. Each step is a structure with
+  fields `Name` (string) and `Args` (a reference to a secret).
+- `DefaultSteps` is an enumeration defaulting to `Enabled`. If `Disabled`,
+  the execution of the default deployment procedure will be supressed.
+- The `Image.URL` field will become optional if `DefaultSteps` is `Disabled`
+  since we don't know if the deploy method actually requires a separate image.
+
+Setting `CustomDeploy.Steps` to a non-empty value and `DefaultSteps` to
+`Disabled` will trigger deployment even if `Image.URL` is empty.
```

Comment:
This seems very... general, and it worries me because it is exposing a _lot_ of ironic internals. I was hoping for something much simpler along the lines of setting something like:

```
  image:
      url: live-iso:<provision_step_name>[:<optional_suffix_that_is_ignored>]
```

and that the existing user/meta/networkData secrets would provide sufficient additional data. Effectively we dictate an API to the author of the custom provision step in IPA, rather than the other way around. In principle that seems like that should be enough for CoreOS, which just needs an ignition file AFAIK?

Perhaps I'm being naïve about how little customisation we can get away with here. We had another proposal (#181) already for a similarly-general interface, although without any use cases provided so it's difficult to evaluate. I'm willing to be educated on this topic, but my strong preference is to expose this functionality in a much more circumscribed way.

Reply:
Well, a non-standard URL schema is definitely not simpler for any definition of "simple" I'm aware of :) I'd prefer not to mix the notion of custom deploy and the image. A custom deploy *may* have an image (actually, even coreos-installer supports using a different image). If we want this to be *actually* usable for anything except for CoreOS, we need to allow at least some degree of flexibility.

I guess you main concern is stemming from providing arbitrary arguments? What I can do is to create one simple field `CustomDeploy` that will unconditionally override the deploy process. This makes this proposal twice less generic (and thus twice less usable FWIW).

Reply:
> Well, a non-standard URL schema is definitely not simpler for any definition of "simple" I'm aware of :)

It's matching a fixed string prefix in an existing API field, difficult to see how it could be much simpler tbh.

> I'd prefer not to mix the notion of custom deploy and the image. A custom deploy _may_ have an image (actually, even coreos-installer supports using a different image). If we want this to be _actually_ usable for anything except for CoreOS, we need to allow at least some degree of flexibility.

What advantage is there to using coreos-installer to install a _different_ image to the one in the ISO, as opposed to just installing the image we want through the regular clean steps?

> I guess you main concern is stemming from providing arbitrary arguments? What I can do is to create one simple field `CustomDeploy` that will unconditionally override the deploy process.

This is definitely an improvement IMHO, thanks.

> This makes this proposal twice less generic (and thus twice less usable FWIW).

Making the most generic solution is not a goal from my perspective, and in fact exposing all of the potential functionality of ironic and all custom deploy steps that could be written is an anti-goal. To my mind the goal here should be to support provisioning methods that don't require a separate provisioning image, without hardcoding the name of any particular technology (e.g. CoreOS) into the API.

Reply:
> It's matching a fixed string prefix in an existing API field, difficult to see how it could be much simpler tbh.

Simpler: a field per feature.

> What advantage is there to using coreos-installer to install a different image to the one in the ISO, as opposed to just installing the image we want through the regular clean steps?

I'm not sure, but they do it for some reason. Non-CoreOS cases may need this as well.